README.md
# dcc104-canvas-exr05-oliveiraleo

-> Controles:

Enter: Inicia a partida

P: Pausa o jogo

R: Reinicia o jogo

-Player 1

Setas: Direcionam a nave

Espaço: Atira

-Player 2

WASD: Direcionam a nave

Ctrl: Atira


changelog.txt
-> Commit 1

- Commit inicial

-> Commit 2

- Ajuste do texto na tela
- Ajuste na posição das naves
- Ajuste na função de tiro

- Correção no tiro do player 2

- Remoção de código desnecessário

-> Commit 3

- Adição do readme

- Ajuste do tamanho das naves (shooters)
- Ajuste da posição inicial das naves (shooters)
- Ajuste do tamanho da tela do jogo
- Ajuste dos estilos (style.css)
- Ajuste dos textos da tela

- Correção da função de tiro
- Correção da função de movimento das naves

- Remoção de controles obsoletos ou duplicados

-> Commit 4

- Adição de uma nova função de tiro
- Adição de novos comentários
- Adição de texto na tela
- Adição de pontuação
- Adição de colisão no tiro
- Adição de colisão das naves com os limites da tela
- Adição do fim de jogo

- Ajuste na função de tiro para deletar os projeteis que sairam da tela
- Ajuste na função de reset do jogo
- Ajuste nas propriedades dos sprites
- Ajuste na função de posição das naves

- Correção dos controles do player 2
- Correção nas funções de velocidade das naves
- Correção na regra do jogo

- Remoção de alguns comentários e códigos obsoletos

-> Commit 5

- Adição de sons
- Adição de novos objetos (no sprite)

- Ajuste nos comentários

- Correção da exibição dos textos na tela
- Correção da função de movimento
- Correção do desenho dos sprites
- Correção na função de tiro

- Remoção de arquivos desnecessários
- Remoção geral de código inutilizado

-> Commit 6

- Ajuste no sprite do player 2

- Correção na movimentação do player 2
- Correção dos limites da tela para as naves (que agora são inerentes ao tipo do objeto e não a instância do mesmo)

- Remoção da barra de energia (desnecessária)


sprite.js
//matematica
const PI2     = Math.PI / 2.0;
const PI3			= Math.PI / 3.0;
const PI4     = Math.PI / 4.0;
const PI6	  = Math.PI / 6.0;
const PI12	  = Math.PI / 12.0;
const RAD4DEG = Math.PI / 180.0;
const DEG4RAD = 180.0 / Math.PI;
//tiros
function Shot(x, y, vx, vy, r, dir, angle) {
	this.pos = {x, y};
	this.vel = {vx, vy};
	this.radius = r; //raio da circunferencia
	this.angle = angle; //angulo do movimento do tiro
	//desenha os tiros
	this.draw = function(ctx) {
		ctx.fillStyle   = "DarkSlateGrey ";
		ctx.strokeStyle = "black";
		//desenha a circunferencia
		ctx.beginPath();
			ctx.arc(this.pos.x, this.pos.y, this.radius, 0, 2 * Math.PI);// coordenadas, raio, ang inicial, ang final (arco)
		ctx.closePath();
		ctx.fill();
	}
	//Verifica se o tiro saiu da tela
	this.isForaTela = function(){
		if(this.pos.y < 0 || this.pos.x < 0 || this.pos.x > 500 || this.pos.y > 500){
			return true;
		}
		return false;
	}
	//move os tiros
	this.movexUp = function(dt){
		if (!(this.isForaTela())) {
			this.pos.y -= 300 * dt;
		}
	}

	this.movexDown = function(dt){
		if (!(this.isForaTela())) {
			this.pos.y += 300 * dt;
		}
	}

	this.moveyDir = function(dt){
		if (!(this.isForaTela())) {
			this.pos.x += 300 * dt;
		}
	}

	this.moveyEsq = function(dt){
		if (!(this.isForaTela())) {
			this.pos.x -= 300 * dt;
		}
	}
	//define a direcao do movimento do tiro
	this.move = function(dt){
		if (this.angle == 90 || this.angle == -270) {
			this.movexUp(dt);
		}if (this.angle == -90 || this.angle == 270) {
			this.movexDown(dt);
		}if (this.angle == 180 || this.angle == -180) {
			this.moveyDir(dt);
		}if (this.angle == 0) {
			this.moveyEsq(dt);
		}
	}
}
//naves
function Shooter(center, size, color, rotacao) {
	this.center = center || {x: 0, y: 0};
	this.size  = size || {w: 50, h: 50};
	this.theta = 0;
	this.omega = 0;
	this.g = 0;
  this.vx = 0;
  this.vy = 0;
	this.color = " ";
	this.pontos = 0;
	this.ax = 0;
	this.ay = 0;
	this.am = 0; //aceleracao de movimento
	this.angle = 90;

	this.ballPos = {x: this.center.x, y: this.center.y - this.size.h / 2};
	//desenha a nave
	this.draw = function(ctx) {
		if(!ctx) return;// somente desenha se existe contexto

		ctx.save();
		ctx.translate(this.center.x, this.center.y);
		ctx.rotate(this.angle*2*Math.PI/360);
		//define a cor
		ctx.fillStyle = color;
		ctx.strokeStyle = "#00ff26";
		ctx.beginPath();//desenha o shooter
			ctx.moveTo(this.size.w / 2, this.size.h / 2);
			ctx.lineTo(-this.size.w / 2,  this.size.h / 32);
			ctx.lineTo(this.size.w / 2, -this.size.h / 2);
		ctx.closePath();
		ctx.fill();
		ctx.stroke();
		//mostra o contorno da caixa de colisao
		if(true){
	    ctx.strokeStyle = "grey";
	    ctx.strokeRect(-this.size.w/2, -this.size.h/2, this.size.w, this.size.h);
	  }
	ctx.restore();
}
	//move a nave
	this.move = function(dt) {
		//verifica os limites da tela
		if (this.center.x < this.size.w/2) {
			this.center.x = this.size.w/2;
		}if (this.center.x > 500-this.size.w/2) {
			this.center.x = 500-this.size.w/2;
		}if (this.center.y < this.size.h/2) {
			this.center.y = this.size.h/2;
		}if (this.center.y > 500-this.size.h/2) {
			this.center.y = 500-this.size.h/2;
		}
		if(this.angle >= 360 || this.angle <= -360){
			this.angle = 0;
		}
		this.vx = this.am*Math.cos(Math.PI*this.angle/180);
		this.vy = this.am*Math.sin(Math.PI*this.angle/180);
		this.vy = this.vy + this.ay*dt;
		this.vx = this.vx + this.ax*dt;
		this.center.x += this.vx * dt;
    this.center.y += this.vy * dt;
	}
  //reset da nave
	this.reset = function() {
		//reseta os contadores
		this.life = 5;
		//reposiciona as naves
		this.reposiciona();
	}

	this.reposiciona = function(){
		if(rotacao == 0){
				this.center = {x:500*Math.random() ,y:30 }
		}else {
			this.center = {x:500*Math.random() ,y:470 }
		}
	}
}

function Barreira(pos, size){
	this.pos = pos;
	this.size = size;

	this.draw = function(ctx){
		ctx.rect(this.pos.x, this.pos.y, this.size.w, this.size.h);
		ctx.strokeStyle = "black";
		//ctx.lineWidth = "2";
		ctx.fillStyle = "red";
		ctx.fillRect(this.pos.x, this.pos.y, this.size.w, this.size.h);
		ctx.stroke();
		//ctx.lineWidth = "0";
	}
}


sprite2.js
//matematica
const PI2     = Math.PI / 2.0;
const PI3			= Math.PI / 3.0;
const PI4     = Math.PI / 4.0;
const PI6	  = Math.PI / 6.0;
const PI12	  = Math.PI / 12.0;
const RAD4DEG = Math.PI / 180.0;
const DEG4RAD = 180.0 / Math.PI;

//tiros
function Shot(_x, _y, _vx, _vy, r, dir) {
	//this.pos = {x: _x, y: _y};
	this.pos = {_x, _y};
	//this.vel = {vx: _vx, vy: _vy};
	this.vel = {_vx, _vy};
	this.radius = r;

	this.draw = function(ctx) {
		ctx.fillStyle   = "#000000";
		ctx.strokeStyle = "#ff0000";
		ctx.beginPath();
			ctx.arc(this.pos.x, this.pos.y, 4, 0, 2 * Math.PI, true);
		ctx.closePath();
		ctx.fill();
	}

	this.move = function(dt, g) {
		//var prevy = this.pos.y;

		if(dir == 0){
			this.pos.x += this.vel.vx * dt;
			//this.pos.y += this.vel.vx * dt - (g/2) * dt * dt;
			this.pos.y += this.vel.vy * dt;
		}else if (dir == 1) {
			this.pos.x += this.vel.vx * dt;
			//this.pos.y += this.vel.vx * dt - (g/2) * dt * dt;
			this.pos.y += this.vel.vy * dt;
		}


		//this.radius -= 0.1;
	}
	this.setVelocityVector = function(o, _mag) {
		//var mag = _mag || 325;
		//var mag = 0;
		if(dir == 0){
			var mag = _mag || 325;
		}if (dir == 1) {
			var mag = _mag || 325;
		}

		var d = this.pos;
		var norm = Math.sqrt( Math.pow(d.x - o.x, 2) + Math.pow(o.y - d.y, 2) );

		this.vel = {vx: (d.x - o.x)/norm, vy: (d.y - o.y)/norm};
		this.vel.vx *= mag;
		this.vel.vy *= mag;
	}

	//this.
}



//naves
function Shooter(center, size, color, rotacao) {
	this.center = center || {x: 0, y: 0};
	this.size  = size || {w: 50, h: 50};
	this.theta = 0;
	this.omega = 0;
  //this.newx = 0;
  //this.newy = 0;

	//this.x = 100;
  //this.y = 100;
  //this.w = 20;
  //this.h = 20;
  this.vx = 0;
  this.vy = 0;
  //this.ax = 0;
  //this.ay = 0;
  //this.ang = 0;
  //this.vang = 0;
  //this.acel = 0;
	this.color = " ";
  //this.cor = "grey";
	this.rotacao = rotacao;
	this.pontos = 0;

	this.ballPos = {x: this.center.x, y: this.center.y - this.size.h / 2};

	this.draw = function(ctx) {
		if(!ctx) return;

		ctx.save();
		ctx.translate(this.center.x, this.center.y);
		//ctx.rotate(this.theta);
		ctx.rotate(rotacao);
		//cor
		//ctx.fillStyle = "#000000";
		ctx.fillStyle = color;
		ctx.strokeStyle = "#00ff26";
		ctx.beginPath();
			ctx.moveTo(-this.size.w / 2, this.size.h / 2);
			ctx.lineTo(this.size.w / 2,  this.size.h / 2);
			ctx.lineTo(0, -this.size.h / 2);
		ctx.closePath();
		ctx.fill();
		ctx.stroke();

		ctx.restore();
	}

	this.move = function(dt, g) {
		//limite de rotacao
		//this.theta += this.omega * dt;
		/*if(this.theta < -PI3) {
			this.theta = -PI3;
		} else if(PI3 < this.theta) {
			this.theta = PI3;
		}*/

		this.ballPos.x = this.center.x + (this.size.h / 2) * Math.sin(this.theta);
		this.ballPos.y = this.center.y - (this.size.h / 2) * Math.cos(this.theta);
    //console.log(this.center);
    this.center.x += this.vx * dt;
    this.center.y += this.vy * dt;

		//new

		//this.vx += this.ax*dt;
    //this.vy += this.ay*dt;

    //this.x += this.vx*dt;
    //this.y += this.vy*dt;
	}
 //var hit = new Audio('sound/siren.mp3');
	/*this.colidiu = function(ball) {
		var p = ball.getFrontPoint();
		var w2 = this.size.w / 2;
		console.log('Tiro');
		if(this.center.x - w2 -10 <= p.x && p.x <= this.center.x + w2 + 10) {
			if(this.ballPos.y + 10 < p.y) {
				this.life -= 1;
				hit.play();
				if(this.life == 0){
					return 2;
				}
				return 1;
			}
		}
		return 0;
	}*/

	this.reset = function() {
		this.life = 3;
		//this.theta = 0;
		//this.omega = 0;
		this.pontos = 0;
	}
}

/*function CollectionGenerator(W, H) {
	this.WIDTH  = W;
	this.HEIGHT = H;

	this.build = function(n){
		var builds = [];
		var lastw = 0;

		for(var i = 0; i < n; i++) {
			var wi = Math.floor(50 + Math.random() * 10);
			var hi = Math.floor(100 + Math.random() * 40);
			var xi = 5 + i * (lastw + 20);
			var yi = this.HEIGHT - hi;


			lastw = wi;

			builds.push(new Build(xi, yi, wi, hi));
		}
		return builds;
	},
	this.asteroid = function(n) {
		var asteroids = [];
		var s = function() {
			return Math.pow(-1, Math.floor(Math.random() * 2));
		}

		for(var i = 0; i < n; i++) {
			var r  = Math.floor(8 + Math.random() * 4);
			var cx = Math.floor(10 + Math.random() * this.WIDTH-10);
			var cy = -r * (i + 2);
			var vx = s() * Math.floor(10 + Math.random() * 10);
			var vy = s() * Math.floor(10 + Math.random() * 10);

			asteroids.push(new Asteroid(cx, cy, r, vx, vy));
		}

		return asteroids;
	}
};*/


index.js
//Texto do placar
function Text(font, size, rgb) {
	this.font = font 	|| "Courier";
	this.size = size 	|| 20;
	this.color = rgb 	|| "#000000" ;

	this.raster = function(ctx, text, x, y) {
		ctx.font = "" + this.size + "px " + this.font;
		ctx.fillStyle = this.color;
		ctx.fillText(text, x, y);// escreve na tela
	}
}
//variaveis globais
var pause = false;
var inicio = false;

var texto = new Text();
var texto2 = new Text("Courier", 20, "blue");

var ganhador = 0;

//sons
var musica = new Audio("sound/war.m4a");

var boom = new Audio();
boom.src = "sound/boom.mp3";

var explode = function(){
	boom.load();
	boom.play();
}

var tiro1 = new Audio();
tiro1.src = "sound/tiro.mp3";
var tiro2 = new Audio();
tiro2.src = "sound/tiro_forte.mp3";

var atira1 = function(){
	tiro1.load();
	tiro1.volume = 1.0;
	tiro1.play();
}

var atira2 = function(){
	tiro2.volume = 0.2;
	tiro2.load();
	tiro2.play();
}
//colisao do shooter com as barreiras
var colideBarreira = function(shooter, parede){
		if(shooter.center.x+(shooter.size.w/2) >= parede.pos.x &&
			shooter.center.x-(shooter.size.w/2) <= parede.pos.x+parede.size.w &&
			shooter.center.y+(shooter.size.h/2) >= parede.pos.y &&
			shooter.center.y-(shooter.size.h/2) <= parede.pos.y+parede.size.h){
		shooter.life--;
		shooter.reposiciona();
		explode();
	}
}
//colisao do tiro com as barreiras
var colideTiroParede = function(tiro, parede){
	if(tiro.pos.x+(tiro.radius/2) >= parede.pos.x &&
		tiro.pos.x-(tiro.radius/2) <= parede.pos.x+parede.size.w &&
		tiro.pos.y+(tiro.radius/2) >= parede.pos.y &&
		tiro.pos.y-(tiro.radius/2) <= parede.pos.y+parede.size.h){
			return true;
		}
	return false;
}
//colisao entre naves
var colideNave = function(s1, s2){
	if(s1.center.x+(s1.size.w/2) >= s2.center.x-(s2.size.w/2) &&
		s1.center.x-(s1.size.w/2) <= s2.center.x+(s2.size.w/2) &&
		s1.center.y+(s1.size.h/2) >= s2.center.y-(s2.size.h/2) &&
		s1.center.y-(s1.size.h/2) <= s2.center.y+(s2.size.h/2)){
			s1.reposiciona();
			s1.life--;
			s2.reposiciona();
			s2.life--;
		}
}

var colideTiros = function(nave, tiro){
	if(nave.center.x+(nave.size.w/2) >= tiro.pos.x &&
		nave.center.x-(nave.size.w/2) <= tiro.pos.x &&
		nave.center.y+(nave.size.h/2) >= tiro.pos.y &&
		nave.center.y-(nave.size.h/2) <= tiro.pos.y){
			nave.life--;
			return true;
		}
	return false;
}
//Regra do jogo
function start() {
	var canvas = document.getElementById("game");
	var ctx = canvas.getContext("2d");// cria o contexto
	// define as constantes
	const WIDTH = canvas.offsetWidth;
	const HEIGHT = canvas.offsetHeight;

	const FPS = 60;
	const DT = 1/FPS;
	const G = -20;

	const PTSMAX = 2; // pontuacao que encerra o jogo
	//variaveis globais
	var shots = [];
	var shoot = false;
	var shots2 = [];
	var shoot2 = false;

	var shooter1 = new Shooter({x: 0, y: 0}, {w: 25, h: 25}, "black", 0);
  var shooter2 = new Shooter({x: 0, y: 0}, {w: 25, h: 25}, "blue", 2*Math.PI);
	//var ball = new Shot(shooter1.ballPos.x, (shooter1.ballPos.y), 0, 0, 12, 0);
  //var ball2 = new Shot(shooter2.ballPos.x, shooter2.ballPos.y, 0, -325, 12, 1);
	//Instancia os obstaculos
	var parede1 = new Barreira({x: WIDTH/6, y: HEIGHT/5}, {w: 50, h: 50});
	var parede2 = new Barreira({x: WIDTH-(WIDTH/3), y: HEIGHT/6}, {w: 50, h: 50});
	var parede3 = new Barreira({x: WIDTH/6, y: HEIGHT-(HEIGHT/3)}, {w: 50, h: 50});
	var parede4 = new Barreira({x: WIDTH-(WIDTH/3), y: HEIGHT-(HEIGHT/4)}, {w: 50, h: 50});
	var parede5 = new Barreira({x: (WIDTH/2)-80, y: (HEIGHT/2)-30}, {w: 250, h: 50});

	var verificaPontos1 = false;
	var verificaPontos2 = false;

	var recomeca = true;
	var verificaInicio = false;

	var msgInicio = new Text("Courier", 30, "black");
	var msg = new Text("Courier", 25, "black");

	//reset do jogo
	function reset() {
		ctx.clearRect(0, 0, WIDTH, HEIGHT);// limpa a tela do jogo
		// limpa os tiros da tela
		shots.length = 0;
		shots2.length = 0;
		if(!recomeca){
			if (verificaInicio) {
				if (ganhador == 1){
					msg.raster(ctx, "Player 1 ganhou a rodada!", WIDTH/8, HEIGHT/4);
				}if (ganhador == 2){
					msg.raster(ctx, "Player 2 ganhou a rodada!", WIDTH/8, HEIGHT-HEIGHT/3);
				}if (shooter1.pontos < PTSMAX && shooter2.pontos < PTSMAX){
					msg.raster(ctx, "Apertem R para continuar", WIDTH/6, HEIGHT/2 );
				}else {
					msg.raster(ctx, "Fim de jogo", WIDTH/3, HEIGHT/2 );
				}
			}
		}
		verificaInicio = true;

		shooter1.reset();//volta as propriedades do shooter ao padrao do inicio
    shooter2.reset();//volta as propriedades do shooter ao padrao do inicio
		ganhador = 0;//reseta o ganhador
	}; reset();
	//regra do jogo
	var loop = function() {
		if(inicio && !pause && recomeca){
		ctx.clearRect(0, 0, WIDTH, HEIGHT);
		//texto da tela do jogo
		texto.raster(ctx, "Player 1", 17, 20);
		texto.raster(ctx, "Pontos: " + shooter1.pontos, 10, 40);
		texto.raster(ctx, "Energia:" + shooter1.life, 10, 60);

		texto2.raster(ctx, "Player 2", WIDTH-103, 20);
		texto2.raster(ctx, "Pontos: " + shooter2.pontos, WIDTH-110, 40);
		texto2.raster(ctx, "Energia:" + shooter2.life, WIDTH-110, 60);
		//popula o vetor de paredes
		var paredes = [];
		paredes.push(parede1);
		paredes.push(parede2);
		paredes.push(parede3);
		paredes.push(parede4);
		paredes.push(parede5);

		var cont, cont2;//contadores para o for
		//tiros player 1
		for(cont = 0; cont < shots.length; cont++) {
			shots[cont].move(DT);
			//Apaga os tiros que bateram em algo
			if(shots[cont].pos.y < 0 ||
				shots[cont].pos.x < 0 ||
				shots[cont].pos.x > WIDTH ||
				shots[cont].pos.y > HEIGHT ||
				colideTiroParede(shots[cont], paredes[0]) ||
				colideTiroParede(shots[cont], paredes[1]) ||
				colideTiroParede(shots[cont], paredes[2]) ||
				colideTiroParede(shots[cont], paredes[3]) ||
				colideTiroParede(shots[cont], paredes[4]) ||
				colideTiros(shooter2, shots[cont])) {// impõe limites no mapa
				shots.splice(cont, 1);// remove o tiro do vetor
			}
		}
		//tiros player 2
		for(cont2 = 0; cont2 < shots2.length; cont2++) {
			shots2[cont2].move(DT);
			//Apaga os tiros que saem da tela
			if(shots2[cont2].pos.y < 0 ||
				shots2[cont2].pos.x < 0 ||
				shots2[cont2].pos.x > WIDTH ||
				shots2[cont2].pos.y > HEIGHT ||
				colideTiroParede(shots2[cont2], paredes[0]) ||
				colideTiroParede(shots2[cont2], paredes[1]) ||
				colideTiroParede(shots2[cont2], paredes[2]) ||
				colideTiroParede(shots2[cont2], paredes[3]) ||
				colideTiroParede(shots2[cont2], paredes[4]) ||
				colideTiros(shooter1, shots2[cont2])) {
				shots2.splice(cont2, 1);// remove o tiro do vetor
			}
		}
		//Movimenta as naves
		shooter1.move(DT);
		shooter2.move(DT);
		//desenha os tiros na tela
		shots.forEach( function(shot) { shot.draw(ctx); } );
		shots2.forEach( function(shot2) { shot2.draw(ctx); } );
		//desenha as naves na tela
		shooter1.draw(ctx);
		shooter2.draw(ctx);
		//desenha os obstaculos
		for(cont = 0; cont < paredes.length; cont++){
			paredes[cont].draw(ctx);
		}
		//verifica a colisao das naves com os obstaculos
		for(cont = 0; cont < paredes.length; cont++){
			colideBarreira(shooter1, paredes[cont]);
		}
		for(cont2 = 0; cont2 < paredes.length; cont2++){
			colideBarreira(shooter2, paredes[cont2]);
		}
		//verifica a colisao entre as naves
		colideNave(shooter1, shooter2);
		//toca a musica de fundo em loop
		musica.volume = 1.0;
		musica.play();
		musica.addEventListener('ended', function() {
			this.currentTime = 0;
			this.play();
		});
		//determina o fim do jogo
		if (shooter1.pontos >= PTSMAX || shooter2.life == 0) {
			ganhador = 1;
			shooter1.pontos++;
			recomeca = false;
			reset();
		}if (shooter2.pontos >= PTSMAX || shooter1.life == 0) {
			ganhador = 2;
			shooter2.pontos++;
			recomeca = false;
			reset();
		}
	}else if(!inicio){// exibe a mensagem da tela inicial
		ctx.clearRect(0, 0, WIDTH, HEIGHT);
		msgInicio.raster(ctx, "Apertem ENTER para começar", 25, HEIGHT/2 );
	}else if(pause){// exibe a mensagem de jogo pausado
		musica.pause();
		msg.raster(ctx, "Apertem P para continuar", (WIDTH/6), HEIGHT/2-parede5.size.h );
	}
}

	setInterval(loop, 1000/FPS);
	//controles do jogo
	addEventListener("keydown", function(e){
		if(e.keyCode == 32 && !shoot) { // Espaco, tiro player 1
			var ball = new Shot(shooter1.ballPos.x, (shooter1.ballPos.y), 0, 0, 4, 0, shooter1.angle);
			if (shooter1.angle == 90 || shooter1.angle == -270) {
				ball.pos = {x: shooter1.center.x, y: (shooter1.center.y)-shooter1.size.h/2};
			}if (shooter1.angle == -90 || shooter1.angle == 270) {
				ball.pos = {x: shooter1.center.x, y: (shooter1.center.y)+shooter1.size.h/2};
			}if (shooter1.angle == 180 || shooter1.angle == -180) {
				ball.pos = {x: shooter1.center.x+shooter1.size.w/2, y: shooter1.center.y};
			}if (shooter1.angle == 0) {
				ball.pos = {x: shooter1.center.x-shooter1.size.w/2, y: shooter1.center.y};
			}
			shots.push(ball); // adiciona a bala no vetor de tiros
			ball = null; // apaga a bala auxiliar
			shoot = true;// bloqueia a repeticao do tiro
			atira1();
			e.preventDefault();
		}if(e.keyCode == 37){ // esquerda player 1
			shooter1.angle -= 90;
			e.preventDefault();
		}if(e.keyCode == 39){ // direita player 1
			shooter1.angle += 90;
			e.preventDefault();
		}if (e.keyCode == 38) { // cima player 1
			shooter1.am = -100;
			e.preventDefault();
    }if (e.keyCode == 40) { // baixo player 1
			shooter1.am = 100;
			e.preventDefault();
    }if(e.keyCode == 13){// Enter
			inicio = true;
			e.preventDefault();
		}if(e.keyCode == 80){// P
				pause = !pause;
		}if (e.keyCode == 82) {// R
				recomeca = true;
		}if (e.keyCode == 87) {// W
			shooter2.am = -100;
			e.preventDefault();
		}if (e.keyCode == 83) {// S
			shooter2.am = 100;
			e.preventDefault();
		}if (e.keyCode == 65) {// A
			shooter2.angle -= 90;
			e.preventDefault();
		}if (e.keyCode == 68) {// D
			shooter2.angle += 90;
			e.preventDefault();
		}if (e.keyCode == 16){// Shift Esq
			var ball2 = new Shot(shooter2.ballPos.x, (shooter2.ballPos.y), 0, 0, 4, 2*Math.PI, shooter2.angle);
			if (shooter2.angle == 90 || shooter2.angle == -270) {
				ball2.pos = {x: shooter2.center.x, y: (shooter2.center.y)-shooter2.size.h/2};
			}if (shooter2.angle == -90 || shooter2.angle == 270) {
				ball2.pos = {x: shooter2.center.x, y: (shooter2.center.y)+shooter2.size.h/2};
			}if (shooter2.angle == 180 || shooter2.angle == -180) {
				ball2.pos = {x: shooter2.center.x+shooter2.size.w/2, y: shooter2.center.y};
			}if (shooter2.angle == 0) {
				ball2.pos = {x: shooter2.center.x-shooter2.size.w/2, y: shooter2.center.y};
			}
			shots2.push(ball2); // adiciona a bala no vetor de tiros
			ball2 = null; // apaga a bala auxiliar
			shoot2 = true;// bloqueia a repeticao do tiro
			atira2();
			e.preventDefault();
		}
	});

	addEventListener("keyup", function(e){
		if(e.keyCode == 32) { // Espaco player 1
			shoot = false;
		}if(e.keyCode == 37 || e.keyCode == 39){ //esquerda e direita player 1
			shooter1.vang = 0;
		}if (e.keyCode == 38 || e.keyCode == 40) { //cima e baixo player 1
			shooter1.am = 0;
    }if (e.keyCode == 87 || e.keyCode == 83) {// W e S
			shooter2.am = 0;
		}if (e.keyCode == 65 || e.keyCode == 68) {// A e D
			shooter2.vang = 0;
		}if (e.keyCode == 16) {// Shift Esq
			shoot2 = false;
		}
	});
}
